File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    l_coordinate, r_coordinate = [3, 0], [3, 2]
    
    global keypad 
    keypad = [[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9], 
              ['*', 0, '#']]
    
    for i in numbers:
        if i in [1, 4, 7]:
            answer += 'L'
        elif i in [3, 6, 9]:
            answer += 'R'
        else:
            l_distance, r_distance = get_distance(i, l_coordinate), get_distance(i, r_coordinate)
            if l_distance == r_distance:
                if hand == 'left':
                    answer += 'L'
                else:
                    answer += 'R'
            elif l_distance > r_distance:
                answer += 'R'
            else:
                answer += 'L'
                
        if answer[-1] == 'L':
            l_coordinate = get_coordinate(i)
        else:
            r_coordinate = get_coordinate(i)
    
    return answer
                
def get_coordinate(n):
    coord = []
    for n_col, key_col in enumerate(keypad):
        if n in key_col:
            coord.append(n_col)
            coord.append(key_col.index(n))
    return coord

def get_distance(n, coordinate):
    n_coord = get_coordinate(n)
    distance = abs(n_coord[0] - coordinate[0]) + abs(n_coord[1] - coordinate[1])
    return distance

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.swift
import Foundation

func solution(_ numbers:[Int], _ hand:String) -> String {
    var result = ""
    let hand = hand == "right" ? "R" : "L"
    var left = (3, 0)
    var right = (3, 2)
    
    for num in numbers {
        let targetCoordinate = getCoordinate(num)
        
        if [1, 4, 7].contains(num) {
            result += "L"
        } else if [3, 6, 9].contains(num) {
            result += "R"
        } else {
            let leftDistance = getDistance(left, targetCoordinate)
            let rightDistance = getDistance(right, targetCoordinate)
            
            if leftDistance > rightDistance {
                result += "R"
            } else if leftDistance < rightDistance {
                result += "L"
            } else {
                result += hand
            }
        }
        
        if let lastResult = result.last {
            if lastResult == "R" {
                right = targetCoordinate
            } else {
                left = targetCoordinate
            }
        }
    }
    return result
}

func getCoordinate(_ num: Int) -> (Int, Int) {
    let coordinateDict = [
        1: (0, 0), 2: (0, 1), 3: (0, 2),
        4: (1, 0), 5: (1, 1), 6: (1, 2),
        7: (2, 0), 8: (2, 1), 9: (2, 2),
        0: (3, 1)
    ]
    
    return coordinateDict[num]!
}
func getDistance(_ coordinate: (Int, Int), _ target: (Int, Int)) -> Int {
    return abs(coordinate.0 - target.0) + abs(coordinate.1 - target.1)
}